Notification frameworks (#162)
neo4j/docs-status-codes: modules/ROOT/pages/notifications/index.adoc

:description: The Neo4j notifications for Neo4j version {neo4j-version}.

[[notifications]]
= Server notifications

> The Neo4j server returns notifications in response to successful query execution to give the users some advice on how to improve the quality of the query or to provide additional information about the query execution.

Author: ```suggestion
The Neo4j server sends notifications after a successful query execution to provide advice on how to improve the query's quality or to provide additional information about the query execution.

Notifications are part of the Neo4j Result API on the server side.
A method called `getNotifications()` returns a list of server-side Notification objects. These notifications are then sent to the driver over Bolt as Success Bolt message metadata.

On the driver side, notifications are part of the ResultSummary API, with a method called `notifications()` that returns a list of driver-side Notification objects.

The result of the `getCode()` or `code()` methods is known as the Neo4j status code.

Driver-side notification configuration filters notifications by severity and/or category at both the driver and session levels.
```

The driver receives these notifications and sends them to the client, which displays them to the user.

From version 5.23, Neo4j has a new GqlStatusObject API in addition to the existing Notification API.
The GqlStatusObject API provides information about the status of a Cypher query or command execution in compliance with the GQL standard.

> This page describes both the Neo4j Notification and the GQL-status object frameworks, how they are structured, the objects they provide for notifications, and how to interpret them.

Reviewer: I think we are still not fully consistent on naming. Should we stick to `GqlStatusObject` (this is what it is called in the codebase) when we talk about the framework/API and `GQL-status object` (this is what it is called in GQL standard) if we talk about the actual object a user get back or what do you think? Can you read through and check that it is consistent?

It also explains the server notification grouping and filtering, the notification internals, and the server-driver compatibility for both the Notification and GqlStatusObject APIs.

[[neo4j-notification-object]]
== Neo4j notification object

The Neo4j notification object contains diagnostic information representing the successful outcome of a Cypher query or command execution, including severity, the `ClientNotification` code, category, title, description, and position in the query text where the notification is relevant.
Depending on the application, some of the fields from the notification object might not be visible.

The notification object consists of the following fields:

.Neo4j notification object
[cols="<1s,<4"]
|===
|Neo4j code
a|The Neo4j status code in the form of `Neo.ClientNotification.[SubType].[Name]`.

> |Title
> a|The title of the notification.
> |Description
> a|The description of the Neo4j status code.

Reviewer: ```suggestion
|Title
a|The title of the Neo4j status code.
|Description
a|The description of the specific notification.
```

Each GQL-status object contains diagnostic information about the execution outcome.
This information includes the GQLSTATUS code, the execution outcome (successful completion, warnings, no data, information), the status description, and the diagnostic record. +
The main GQL-status object, describing the condition with the greatest precedence, is called the primary GQL-status object.
The primary GQL-status object is always present in the execution outcome and follows the GQL precedence rules:

. `NO DATA` has precedence over `WARNING`.
. `WARNING` has precedence over the `SUCCESSFUL COMPLETION` subclass.
. `SUCCESSFUL COMPLETION` subclass has precedence over `INFORMATIONAL`.
. `INFORMATIONAL` is the condition with the least precedence.

The primary GQL-status object also includes the Neo4j-specific information, such as the severity level and the classification of the notification, which can be used for filtering.
For more information about notification grouping and filtering, see <<notification-grouping-and-filtering>>.

The additional GQL-status objects are part of the list of additional GQL-status objects of the execution outcome.

> The list does not follow any specific order and might be empty as well.

Reviewer: This is not how we implement it in Neo4j. We have one list with all GQL-status object (including the primary one). The whole list is ordered after precedence, meaning the primary object is always first. This is not exactly how GQL describes it but it was agreed this would be more useful and could  also be considered compliant.

The GQL-status object consists of the following fields:

.GQLSTATUS notification object
[cols="<1s,<4"]
|===
|GQLSTATUS code
a| A 5-character string that is the concatenation of a 2-character class code followed by a 3-character subclass code.

> |GQLSTATUS description

Reviewer: ```suggestion
|StatusDescription
```

a| A human-readable description of the GQLSTATUS, which consists of a condition, a subcondition, and a description.

> The condition and subcondition are the class and subclass codes, respectively.

Reviewer: ```suggestion
The condition and subcondition represents the class and subclass codes, respectively.
```

Not sure if represents is a good word. Possible you can write out that they are textual representations of the class and subclass codes or something. But the class and subclass codes the 5-digit alphanumeric ones while condition and subcondition is text

| DiagnosticRecord
a| Extra information about the status, given as key-value pairs, both on the server and driver side.

> The full diagnostic record can be returned by the `getDiagnosticRecord()` method on the server side and the `diagnosticRecord()` methods on the driver side.

Reviewer 2: ```suggestion
To retrieve the full diagnostic record, you can use the `diagnosticRecord()` method on both the server and driver sides.
```

Reviewer: Good suggestion, please go ahead and merge

> The full diagnostic record can be returned by the `getDiagnosticRecord()` method on the server side and the `diagnosticRecord()` methods on the driver side.

Reviewer: We ended up calling it just `diagnosticRecord()` on the server side as well

Additional helper methods are exposed for some useful fields.
[cols="<1s,<4"]
!===
! Field ! Description
! OPERATION ! The operation that the notification is related to.
! OPERATION_CODE ! The operation code that the notification is related to.

> ! CURRENT_SCHEMA ! The current schema that the notification is related to.

Reviewer: We might want to add a note saying that we only have these because GQL requires it and that they will always just give a default value and therefore are pointless. You might want to phrase that in a nicer way though xD

Author: Do we want them documented, then?

Author: I'm a bit wary to say that we don't use them.

Reviewer: I think we have to in order to be compliant. I think what you wrote now looks good

> ! severity ! The Neo4j severity level, which can be one of the following:

Reviewer: It shuld be _severity, _classification, _position and _status_parameters all starting with an underscore

The class code indicates the general condition of the status (such as successful completion, warning, or information), and the subclass code provides more detailed information about the condition, such as classification and messages.

The following table lists the Neo4j-defined groups of GQLSTATUS codes and their meanings:

.GQLSTATUS groups of codes as defined by Neo4j
[frame="topbot", stripes=odd, grid="cols", cols="<1s,<4"]
|===
|GQLSTATUS code
|*Description*

| 01N0[y]
| Deprecation warnings

| 01N3[y]
| Hint warnings

| 01N4[y]
| Unsupported warnings

| 01N5[y]
| Unrecognized warnings

| 01N6[y]
| Generic warnings

| 01N7[y]
| Security warnings

| 03N9[y]
| Performance information

| 03N6[y]
| Generic information

| 00N5[y]
| Unrecognized information under successful completion

| 00N6[y]
| Generic information under successful completion

| 00N7[y]
| Security information under successful completion

| 00N8[y]
| Topology information under successful completion

|===

[[notification-grouping-and-filtering]]
== Server notification grouping and filtering

> All server notifications are grouped by category (which is called classification in the GQLSTATUS framework) and severity level, which can be one of `WARNING`, `WARNING OR INFORMATION`, or `INFORMATION`.

Reviewer: ```suggestion
All server notifications are grouped by category (which is called classification in the GqlStatusObject framework) and severity level, which can be one of `WARNING`, `WARNING OR INFORMATION`, or `INFORMATION`.
```

> The driver-side notification configuration used for filtering notifications by category and severity is the same for both Neo4j and GQLSTATUS notification frameworks.

Reviewer: ```suggestion
The driver-side notification configuration used for filtering notifications by category and severity is the same for both Neo4j Notification and GQL-status object frameworks.
```

> The driver-side notification configuration used for filtering notifications by category and severity is the same for both Neo4j and GQLSTATUS notifications.

Reviewer: As above I don't like the naming 'Neo4j and GQLSTATUS notifications' for the old and new frameworks

| INFORMATION
| The query uses costly operations and might be slow. Consider if it is possible to rewrite the query in a different way.
|
|===

[[server-driver-compatibility]]
== Server-driver compatibility

The server and driver communicate with each other through the Bolt protocol.
During the handshake process, they agree on using the newest possible Bolt protocol version that both the server and the driver support.
For more information on the Bolt versions supported by different server versions, see the link:https://neo4j.com/docs/bolt/current/bolt-compatibility[Bolt Protocol documentation].

The GqlStatusObject API is available in Neo4j 5.22 and later versions on the server side and in the 5.23 driver and later versions on the driver side.
The current Notification API is still present, and the GqlStatusObject API can be used in parallel with it.

> To receive the Gql-status notifications, both your server and the driver must support the GqlStatusObject API.

Reviewer: ```suggestion
To use the GqlStatusObject API, both your server and the driver must support it.
```

Although that is not technically true as you can use it 
1. directly through the server API
2. using a new driver and an old server and let the driver polyfill as described below.

I think the vast majority of our users use Neo4j with a driver though and scenario 2 is not super helpful in comparison with if the server was newer. Maybe we can phrase it like you must have support on both sides to fully utilize the API or something?
